src/content/docs/references/docs/optionals.md

```diff
@@ -5,187 +5,272 @@ sidebar:
+### What is a `fault`?
 
-1. A variable that works as a union between a `fault` value and an actual value, we call this latter the "result".
-2. A return value that is either a `fault` or a result.
-
-We can construct an Optional by adding the `!` suffix to a type:
-
-```c
-int! a; // An optional value with the result type of "int"
-
-fn int! foo() // A function returning an optional value with the result type of "int"
+A `fault` represents a reason why a `result` could not be returned, like opening a file but finding it does not exist and returning an `IoError.FILE_NOT_FOUND` for instance. A `fault` is defined similarly to simple enums:
+```c3
+fault NoPurchaseReason
 {
-  ...
-}  
-```
-
-It is not possible to create an Optional of an Optional (so for example `int!!` is never valid).
+    BOOK_NOT_FOUND,
+    BOOK_OUT_OF_STOCK,
+}
 
-*Using* a variable or return with an Optional type yields an optional:
+fn void! main(String[] args)
+{
+    NoPurchaseReason order_failed = NoPurchaseReason.BOOK_NOT_FOUND;
 
-    int! a = ...
-    fn int! foo() { ... }
-    
-    double! x = a * 3.14159;
-    double! y = foo() * 0.3333333333;
-    double! z = x * y;
+    // The union of all of the program's `fault` types is an `anyfault` type
+    // You can convert between different types of `fault`
+    anyfault example = order_failed;
+    example = IoError.FILE_NOT_FOUND; // Also ok
+}
+```
 
-Similar to basic operations it's possible to use an Optional value as a call parameter. The return value then becomes Optional
+### What is an Optional?
 
-    int! a = ...
-    fn double bar(int x) { ... }
+Optionals are a tagged union of either the `result` **or** `fault`, either of which can be unpacked in C3.
 
-    // "bar" can be called because the result type of the Optional is "int"
-    // but the return value will become Optional:
-    double! x = bar(a);
+Similar to a "Result" type in other languages, the `fault` is specific and something you can unpack and test against and switch over.
 
-## Optional execution
+Create an Optional from an existing type by appending `!` to that type.
+```c3
+int! example = 6; // Optional value with `result` type of int
+```
 
-The most important feature of Optionals is how it will only optionally execute
-operations. Given a call with the arguments a0, a1, ... the call will only 
-be invoked if all of the arguments evaluate to real values.
+#### Setting the `result` or the `fault` in an Optional
 
-    int! a = 1;
-    int! b = MyResult.FOO?;
+- To set the `result` of an Optional, use regular assignment
+- To set the `fault` `?` suffix operator.
+```c3
+int! a = 1;                       // Set the `result` with `=`
+int! b = IoError.FILE_NOT_FOUND?; // Set the `fault` with '?' 
+```
 
-    // "baz" called, because "a" is has a result. 
-    int! z = baz(a, a);  // same as z = bar(1, 1)
+#### Run code if we find a `fault` in the Optional: 
+```c3
+// Detect `fault` inside "optional_value", handle inside scope
```

Comment:
This is hard. Typically for Optionals we talk about it being "empty" or "contains a value". If we use the word "has a fault", then this is weird: `anyfault! x;` which is quite fine in C3. I think of the C3 optional having 3 things really:

1. It can contain a value, optionally
2. It can be "empty", in which case there isn't a value.
3. IF it is empty it also contains a *reason* why it's empty which is the *fault* (although I have certainly tried finding alternative names to *fault* that would give less association to an error)

These terms are established:
1. When we simply use `if (catch foo)` then we "test it if it's empty" or "test for Empty". 
2. If we do `if (try foo())` we "test if it has a value" or "test for Value".
3. When we use `if (try foo = bar())` we "unwrap the optional"

However, when we have `if (catch f = foo)`, then no terms are established. Maybe it could be like "unwrap the Empty's fault" "get the underlying fault" "get the reason for Empty"

Reply:
terminology agreed:

> testing if a value is missing `if (catch optional_value)`
testing if a value is present  `if (try optional_value)`
unwrapping a value with `if (try value = optional_value)`
retrieve a fault with `if (catch reason = optional_value)`